File: Packages/RallyCore/Sources/RallyCore/Models/Reward.swift
import Foundation

/// Represents a redeemable reward in the loyalty catalog.
public struct Reward: Identifiable, Codable, Hashable, Sendable {
    public let id: String
    public let schoolID: String
    public let title: String
    public let description: String
    public let pointsCost: Int
    public let imageURL: URL?
    public let category: RewardCategory
    public let minimumTier: Tier
    public let sponsorID: String?
    public let inventory: Int?
    public let expiresAt: Date?
    public let isActive: Bool

    public init(
        id: String,
        schoolID: String,
        title: String,
        description: String,
        pointsCost: Int,
        imageURL: URL? = nil,
        category: RewardCategory = .merchandise,
        minimumTier: Tier = .rookie,
        sponsorID: String? = nil,
        inventory: Int? = nil,
        expiresAt: Date? = nil,
        isActive: Bool = true
    ) {
        self.id = id
        self.schoolID = schoolID
        self.title = title
        self.description = description
        self.pointsCost = pointsCost
        self.imageURL = imageURL
        self.category = category
        self.minimumTier = minimumTier
        self.sponsorID = sponsorID
        self.inventory = inventory
        self.expiresAt = expiresAt
        self.isActive = isActive
    }
}

public enum RewardCategory: String, Codable, Hashable, Sendable, CaseIterable {
    case merchandise
    case concessions
    case experiences
    case tickets
    case digital
    case partner
}

/// Represents a reward redemption transaction.
public struct Redemption: Identifiable, Codable, Hashable, Sendable {
    public let id: String
    public let rewardID: String
    public let userID: String
    public let pointsSpent: Int
    public let redeemedAt: Date
    public let status: RedemptionStatus
    public let redemptionCode: String?
    public let expiresAt: Date?

    public init(
        id: String,
        rewardID: String,
        userID: String,
        pointsSpent: Int,
        redeemedAt: Date = .now,
        status: RedemptionStatus = .pending,
        redemptionCode: String? = nil,
        expiresAt: Date? = nil
    ) {
        self.id = id
        self.rewardID = rewardID
        self.userID = userID
        self.pointsSpent = pointsSpent
        self.redeemedAt = redeemedAt
        self.status = status
        self.redemptionCode = redemptionCode
        self.expiresAt = expiresAt
    }
}

public enum RedemptionStatus: String, Codable, Hashable, Sendable {
    case pending
    case confirmed
    case used
    case expired
    case cancelled
}

File: Android/core/src/main/java/com/rally/app/core/model/Reward.kt
package com.rally.app.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents a redeemable reward in the loyalty catalog.
 */
@Serializable
data class Reward(
    val id: String,
    val schoolID: String,
    val title: String,
    val description: String,
    val pointsCost: Int,
    val imageURL: String? = null,
    val category: RewardCategory = RewardCategory.MERCHANDISE,
    val minimumTier: Tier = Tier.ROOKIE,
    val sponsorID: String? = null,
    val inventory: Int? = null,
    val expiresAt: Long? = null,
    val isActive: Boolean = true
)

/**
 * Categories for organizing rewards in the catalog.
 */
@Serializable
enum class RewardCategory {
    @SerialName("merchandise") MERCHANDISE,
    @SerialName("concessions") CONCESSIONS,
    @SerialName("experiences") EXPERIENCES,
    @SerialName("tickets") TICKETS,
    @SerialName("digital") DIGITAL,
    @SerialName("partner") PARTNER;

    val displayName: String
        get() = name.lowercase().replaceFirstChar { it.uppercase() }
}

/**
 * Represents a reward redemption transaction.
 */
@Serializable
data class Redemption(
    val id: String,
    val rewardID: String,
    val userID: String,
    val pointsSpent: Int,
    val redeemedAt: Long = System.currentTimeMillis(),
    val status: RedemptionStatus = RedemptionStatus.PENDING,
    val redemptionCode: String? = null,
    val expiresAt: Long? = null
)

/**
 * Current status of a reward redemption.
 */
@Serializable
enum class RedemptionStatus {
    @SerialName("pending") PENDING,
    @SerialName("confirmed") CONFIRMED,
    @SerialName("used") USED,
    @SerialName("expired") EXPIRED,
    @SerialName("cancelled") CANCELLED
}
